Analysis refactor part 3

zooba opened the pull request:
This is not yet ready to merge, but is getting fairly close. I've done enough of the changes that I want to for this (already huge) PR, so I'm focusing on stabilizing it and getting it merged now.

@MikhailArkhipov With this, you should have enough LSP in the Analysis project to just build that project as a DLL and wrap it up in your own host (that instantiates the `Server` class and passes through messages to it) - shouldn't need any other dependencies. It likely won't be functional for you right away given the hard failure for unexpected messages, but feel free to start making whatever tweaks are needed there to get it running.

Signature help is the next item that needs migrating. I'll do that in the next PR - this one is big enough.

huguesv commented:
After looking at the changes for some time I haven't seen anything objectionable.

I pulled your branch and started testing locally because I figure I have more chances of finding issues that way than staring at various functions with a bunch of "index management" that I need to execute in my head.

First project load was ok, but second gave me this assert:
```
---------------------------
Assertion Failed: Abort=Quit, Retry=Debug, Ignore=Continue
---------------------------
Unhandled exception in Microsoft.PythonTools.Editor.PythonTextBufferInfo.InvokeSinks (C:\Projects\GitHub\zooba\PTVS\Python\Product\PythonTools\PythonTools\Editor\PythonTextBufferInfo.cs:282)

System.NullReferenceException: Object reference not set to an instance of an object.

   at Microsoft.PythonTools.Intellisense.VsProjectAnalyzer.<>c__DisplayClass87_0.<<GetMissingImportsAsync>b__1>d.MoveNext() in C:\Projects\GitHub\zooba\PTVS\Python\Product\PythonTools\PythonTools\Intellisense\ProjectAnalyzer.cs:line 820

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()

   at Microsoft.PythonTools.Intellisense.VsProjectAnalyzer.<EnsureSingleRequest>d__155`2.MoveNext() in C:\Projects\GitHub\zooba\PTVS\Python\Produ......

<truncated>
---------------------------
Abort   Retry   Ignore   
---------------------------
```

huguesv commented:
I've gotten this assert a few times while editing:
```
---------------------------
Assertion Failed: Abort=Quit, Retry=Debug, Ignore=Continue
---------------------------
Buffer content mismatch



   at Microsoft.PythonTools.Intellisense.BufferParser.ValidateBufferContents(IEnumerable`1 snapshots, FileUpdateResponse response) in C:\Projects\GitHub\zooba\PTVS\Python\Product\PythonTools\PythonTools\Intellisense\BufferParser.cs:line 233

   at Microsoft.PythonTools.Intellisense.BufferParser.<ParseBuffersAsync>d__43.MoveNext() in C:\Projects\GitHub\zooba\PTVS\Python\Product\PythonTools\PythonTools\Intellisense\BufferParser.cs:line 279

   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(Object stateMachine)

   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)

   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)

   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()

   at System.Runtime.CompilerServices.AsyncMethodBuilderC......

<truncated>
---------------------------
Abort   Retry   Ignore   
---------------------------
```

huguesv commented:
I've seen refactor-rename work in one scenario, but several others lead to the command not doing anything, with no preview, like this:
![image](https://user-images.githubusercontent.com/1696845/34703664-0492411e-f4a9-11e7-9388-d64f72823944.png)

huguesv commented:
If I start by creation a new application and write some code, I don't get any assert but IntelliSense doesn't give any completions.
If I save that project, close and reload solution, I then get a bunch of mismatch asserts, but IntelliSense seem to work better.
I've tried the "missing import" code fix feature, but it doesn't show up as available, with either of the above scenarios.

huguesv commented:
FYI I tested the latest commits for a bit. Extract method is doing better. I still get the buffer content mismatch (but maybe less frequently) and also the save as dialog.

zooba commented:
@huguesv Awesome, thanks! I think I've fixed the update ordering issue (though we can probably now lose an update completely, so I'll need a custom event for "I'm out of sync send me the whole document again").

Extract Method should be back in action again - I accidentally reversed a type check condition that broke everything. Got a bit more minor cleanup coming, but it's worth grabbing new sources and seeing how it works. (My non-UI test numbers are higher than master, but that's probably because I fixed that invalid header bug. Working on the UI tests today.)

zooba commented:
I think I've got the buffer consistency issues with that last commit (I was grabbing the wrong version number from the message), so now I'll let the tests run through and work on adding more to help validate the LSP interfaces.

huguesv commented:
Sorry, but not quite :( I just took your commits and rebuilt. I opened my existing project and just added a `import sys` and pressed enter at the end of my file and got the assert.

![image](https://user-images.githubusercontent.com/1696845/34752267-0b962e54-f566-11e7-828f-80e73d3171f3.png)

AlexanderSher commented on Python/Product/Analysis/Infrastructure/PathUtils.cs:
```diff
@@ -318,5 +318,87 @@ static class PathUtils {
+        /// <exception cref="ArgumentException">
+        /// If the provided path contains invalid characters.
+        /// </exception>
+        public static string NormalizePath(string path) {
```
We have `Microsoft.PythonTools.Infrastructure.PathUtils.NormalizePath`. How this one is different?

zooba replied:
The Analysis project has no dependencies, so it can't access that.

We can change the existing one to redirect here and only have one implementation (I *think* everything that uses Common also uses Analysis, though it isn't required, and arguably these helpers should be private since the Analysis API is going to be public/stable), but I've mainly just been copy-pasting into the new namespace.

AlexanderSher commented on Python/Product/Analysis/AnalysisLogWriter.cs:
```diff
@@ -222,36 +223,12 @@ public struct LogItem {
             }
             var enc = new UTF8Encoding(false);
-            var stream = OpenRawLogFile(true, true);
+            var stream = PathUtils.OpenWithRetry(_outputFile, FileMode.Append, FileAccess.Write, FileShare.Read);
```
Why are we trying to get file access several times?

zooba replied:
Because we've seen random IO failures in the past that work on retry (presumably AV or disk tools, but it's near impossible to diagnose).

AlexanderSher commented on Python/Product/Analysis/DocumentBuffer.cs:
```diff
@@ -0,0 +1,94 @@
+
+        public void Update(IEnumerable<DocumentChangeSet> changes) {
+            foreach (var change in changes) {
+                Update(change);
```
If one of the updates fails with exception, the others won't be processed. Is that ok?

zooba replied:
Yep. We do need to add an "out of sync" message for the server to raise in this case though (and some others), as there's no way to return failure from a document changed event.

AlexanderSher commented on Python/Product/Analysis/DocumentBuffer.cs:
```diff
@@ -0,0 +1,94 @@
+            }
+
+            int lastStart = int.MaxValue;
+            var lineLoc = SplitLines(Text.ToString()).ToArray();
```
Not sure how critical is this, but allocating new string for indexing is not required.

zooba replied:
I think StringBuilder is missing IndexOf[Any], and I didn't want to rewrite it. I'll mark it as a potential optimization and we can do it in its own PR.

AlexanderSher commented on Python/Product/Analysis/Parsing/Tokenizer.cs:
```diff
@@ -2663,6 +2663,7 @@ public enum NewLineKind {
+    [DebuggerDisplay("NewLineLocation({_endIndex}, {_kind})")]
     public struct NewLineLocation : IComparable<NewLineLocation> {
         private int _endIndex;
         private NewLineKind _kind;
```
These fields should be made `readonly`

AlexanderSher commented on Python/Product/Analysis/DocumentBuffer.cs:
```diff
@@ -0,0 +1,94 @@
+                    continue;
+                }
+
+                int start = NewLineLocation.LocationToIndex(lineLoc, change.ReplacedSpan.Start, Text.Length);
```
Are the `NewLineLocation` instances still valid after `Text` is changed during first iteration?

zooba replied:
Changes have to be ordered in reverse order (according to LSP, but we also check on the next line), so the locations after the change are invalid, but we shouldn't try to use them. They're valid enough for the check to work.

AlexanderSher replied:
Sorry, I don't get it.
Imagine we have string `123456` and two changes: 
- remove range `[1..3]`: `123456 -> 156`
- add text `789` into position `2`: `156 -> 157896`

If they are handled in reverse order, we will get:
- add text `789` into position `2`: `123456 -> 127893456`
- remove range `[1..3]`: `127893456 -> 193456`

AlexanderSher replied:
Also, I think reverse order isn't guaranteed. Changes are reversed in `BufferParser.GetUpdatesForSnapshot`, but in `OutOfProcProjectAnalyzer.UpdateContent` they are simply merged:
```cs
                    changes.AddRange(fileChange.changes.Select(c => new LS.TextDocumentContentChangedEvent {
                        range = new SourceSpan(
                            new SourceLocation(c.startLine, c.startColumn),
                            new SourceLocation(c.endLine, c.endColumn)
                        ),
                        text = c.newText
                    }));
```

zooba replied:
Reverse order of start location, not reverse order of change. I don't recall if there's a non-overlapping requirement, or if it just works out that VS doesn't send overlapping changes in the same update, but this area is *woefully* underspecified and I don't think we have much of a choice other than assuming that our clients (VS/VSC) behave sensibly.

For VS, the changes are guaranteed to be in increasing start location when we get them, so reversing the order to match LSP is fine.

zooba replied:
Within each snapshot, which is why we recalculate the line locations between each one.

AlexanderSher replied:
So `changes.ToVersion - changes.FromVersion == 1`?

zooba replied:
According to the LSP spec, yes. For VS, we expect them to match snapshot version numbers, so the increment is only a fallback and we prefer the explicit value if it's provided. VS Code apparently assumes that updates are synchronous and never dropped.

AlexanderSher replied:
Ok, got it.

huguesv commented:
![image](https://user-images.githubusercontent.com/1696845/34853947-6d9732de-f6eb-11e7-92a0-5f6bcd86d4e6.png)

System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
Parameter name: currentPosition 
at Microsoft.VisualStudio.Text.Tracking.TrackPositionBackwardInTime(PointTrackingMode trackingMode, Int32 currentPosition, ITextVersion currentVersion, ITextVersion targetVersion)
at Microsoft.PythonTools.Intellisense.LocationTracker.Translate(SourceLocation loc, Int32 fromVersion, Int32 toVersion) in C:\Projects\GitHub\zooba\PTVS\Python\Product\PythonTools\PythonTools\Intellisense\LocationTracker.cs:line 269
at Microsoft.PythonTools.PythonAnalysisClassifier.GetClassificationSpans(SnapshotSpan span) in C:\Projects\GitHub\zooba\PTVS\Python\Product\PythonTools\PythonTools\PythonAnalysisClassifier.cs:line 171
at Microsoft.VisualStudio.Text.Classification.Implementation.ClassifierTagger.<GetTags>d__5.MoveNext()
at Microsoft.VisualStudio.Text.Tagging.Implementation.TagAggregator`1.<GetTagsForBuffer>d__39.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)

huguesv commented:
Completions appear when you're defining a function with multiple parameters. I compared with master branch, and that doesn't happen there, so it's a regression from this PR.

![image](https://user-images.githubusercontent.com/1696845/34854139-ab01ad2e-f6ec-11e7-96f7-69a69419bd9b.png)

zooba commented:
Completion triggering is going to change again - LSP embeds all of its logic behind a single message, whereas we use a range of messages depending on context. So that will all have to be reimplemented on the other side anyway.

For now, I'll make sure we don't trigger normal completions on space. That should get us back to at least sensible behaviour, even if it doesn't match the old behaviour 100%.

AlexanderSher commented on Python/Product/Analysis/ExpressionFinder.cs:
```diff
@@ -162,37 +188,29 @@ private class NormalExpressionWalker : ExpressionWalker {
+                return false;
             }
 
             public override bool Walk(Arg node) {
```
So this one always return `false`?

zooba replied:
`false` here means don't continue walking down this branch. In this case, it's because we've already walked it (the `node.Body.Walk(this)` above).

AlexanderSher commented on Python/Product/Analysis/ModulePath.cs:
```diff
@@ -768,6 +776,73 @@ out errorParameter
+            if (isPackage == null) {
+                isPackage = f => !string.IsNullOrEmpty(GetPackageInitPy(f));
+            }
+            if (getModule == null) {
```
`getModule` isn't used anywhere

AlexanderSher commented on Python/Product/Analysis/ProjectEntry.cs:
```diff
@@ -36,34 +39,48 @@ namespace Microsoft.PythonTools.Analysis {
         private ManualResetEventSlim _curWaiter;
         private int _updatesPending, _waiters;
+        private readonly SortedDictionary<int, DocumentBuffer> _buffers;
         internal readonly HashSet<AggregateProjectEntry> _aggregates = new HashSet<AggregateProjectEntry>();
```
Can be made private

zooba replied:
I haven't touched aggregates at all in this PR and I'd rather not start now.

AlexanderSher commented:
>         private void ClearScope() {

Not used

---
Refers to: Python/Product/Analysis/ProjectEntry.cs:279 in 0903354. [](commit_id = 09033543fa7e7a21cb72e36012275dd03ed2b7eb, deletion_comment = False)

huguesv commented:
I'm unable to get completions on project-local packages. In picture below, I get completions on `os`, `class1` but not `package1`. I've tried a number of projects and reloads, but it never worked.
![image](https://user-images.githubusercontent.com/1696845/34892081-ba9c69fe-f78c-11e7-9d8e-7a593eb4ff1e.png)

AlexanderSher commented on Python/Product/Analysis/ProjectEntry.cs:
```diff
@@ -280,60 +283,31 @@ where fileName.StartsWith(pathPrefix)
-        public PythonAst Tree {
-            get { return _tree; }
-        }
+        public PythonAst Tree { get; private set; }
```
This property is accessed under lock inside `ProjectEntry` but is available without lock `ModuleAnalysis` and several other places

zooba replied:
Reading Tree doesn't require the lock - Cookie and _updatesPending do. GetTreeAndCookie reads under the lock to ensure you get matched tree and cookie, but directly accessing tree is not guaranteed to be the latest version.

AlexanderSher commented on Python/Product/Analysis/ProjectEntry.cs:
```diff
@@ -36,34 +39,48 @@ namespace Microsoft.PythonTools.Analysis {
+            FilePath = filePath;
+            Cookie = cookie;
+            MyScope = new ModuleInfo(ModuleName, this, state.Interpreter.CreateModuleContext());
+            _unit = new AnalysisUnit(Tree, MyScope.Scope);
```
I think tree is always `null` here

AlexanderSher commented on Python/Product/Analysis/ProjectEntry.cs:
```diff
@@ -342,15 +316,101 @@ where fileName.StartsWith(pathPrefix)
+                return new MemoryStream(encoding.GetBytes(text.ToString()));
+            }
+
+            var ms = new MemoryStream(encoding == Encoding.Unicode ? text.Length * 2 : text.Length);
```
I'm not sure that `==` is safe. I'd rather use `Equals`.

AlexanderSher commented on Python/Product/Analysis/ProjectEntry.cs:
```diff
@@ -342,15 +316,101 @@ where fileName.StartsWith(pathPrefix)
+            }
+        }
+
+        internal static Stream EncodeToStream(StringBuilder text, Encoding encoding, int chunkSize = 4096) {
```
I think it can be private

zooba replied:
It needs tests (guess that's my job this morning)

AlexanderSher commented on Python/Product/Analysis/ProjectEntry.cs:
```diff
@@ -342,15 +316,101 @@ where fileName.StartsWith(pathPrefix)
             _aggregates.Add(into);
         }
+
+        public IEnumerable<int> DocumentParts => _buffers.Keys.AsLockedEnumerable(_buffers);
```
This is pretty much the same as `lock(this)`. Any client code that calls `DocumentParts.GetEnumerator()` acquires a lock that controls access to private implementation.

zooba replied:
When used correctly (which is to enumerate this sequence and use the values to index into `_buffers`) it's going to be necessary to lock the outer object on all uses. This way, we get a single lock that lasts for the entire `foreach` and prevents races.

Either way requires the caller to use the object correctly. Not disposing an `IDisposable` is more likely to be done correctly than locking the object everywhere.

AlexanderSher commented on Python/Product/Analysis/LanguageServer/ParseQueue.cs:
```diff
@@ -0,0 +1,209 @@
+            if (created == null) {
+                // Do not start parsing until the previous one has completed
+                await tcs.Task;
+                tcs = new TaskCompletionSource<IAnalysisCookie>();
```
If this method is called three times in a row for the same doc, second and third calls will await for the first one and then will run concurrently. You probably need an async ARE here.

AlexanderSher replied:
Or you can simplify this by using `AddOrUpdate`
```cs
public Task<IAnalysisCookie> Enqueue(IDocument doc, PythonLanguageVersion languageVersion) {
    return _parsing.AddOrUpdate(doc, 
        d => Task.Run(() => ParseWorker(d, languageVersion)),
        (d, task) => task.ContinueWith(t => ParseWorker(d, languageVersion)));
}
```
and get rid of the `TaskCompletionSource` altogether.

zooba replied:
Oh that's nice. I didn't know you got the previous value in AddOrUpdate.

huguesv commented:
In one session, I lost all IntelliSense / editor tooltips. It worked again after reloading.
In most sessions, I got many "buffer content mismatch" asserts. For a while, I thought we were rid of those but then I mentioned to Alex that things looked relatively good and it jinxed it.
It looks like the analyzer process doesn't stop when you quit VS? I don't know how frequently that happens.
![image](https://user-images.githubusercontent.com/1696845/34897464-f070cd48-f7a2-11e7-8951-935ca18b51b9.png)

AlexanderSher commented on Python/Product/Analysis/ProjectEntry.cs:
```diff
@@ -153,7 +167,7 @@ internal sealed class ProjectEntry : IPythonProjectEntry, IAggregateableProjectE
 
             lock (this) {
-                _analysisVersion++;
+                AnalysisVersion++;
```
This is the only place where `AnalysisVersion` is accessed under lock

zooba replied:
It's almost the only place where it's modified. I'll lock the other one, but most of the reads are fine.

AlexanderSher commented:
>                 return new StreamReader(new PartiallyReadStream(readBytes, stream), encoding);

`PartiallyReadStream` doesn't override `Dispose`. Who closes the `stream`?

---
Refers to: Python/Product/Analysis/Parsing/Parser.cs:5031 in 0903354. [](commit_id = 09033543fa7e7a21cb72e36012275dd03ed2b7eb, deletion_comment = False)

AlexanderSher commented on Python/Product/Analysis/Infrastructure/ProcessHelper.cs:
```diff
@@ -126,11 +138,12 @@ sealed class ProcessHelper : IDisposable {
+            if (!_process.HasExited) {
+                await _seenNullOutput.WaitAsync(cancellationToken).ConfigureAwait(false);
+                await _seenNullError.WaitAsync(cancellationToken).ConfigureAwait(false);
+                cancellationToken.ThrowIfCancellationRequested();
```
Do you really want to throw OCE when process has completed it's work?

zooba replied:
Semantically, the decision has been made to cancel, which presumably means we also want to cancel any subsequent work as well. Though since we've just completed a cross-process task without cancellation and are not marshalling back to the original thread here, it's very unlikely that a cancellation will occur between the previous wait and this check. (I suspect the call has stuck around from a previous iteration of this code... took a few goes to get `Process` termination right.)

zooba commented:
`PartiallyReadStream` wraps a user argument, so it should be disposed by the original caller.

AlexanderSher commented on Python/Product/Analysis/ProjectEntry.cs:
```diff
@@ -342,15 +320,118 @@ where fileName.StartsWith(pathPrefix)
+                return null;
+            }
+            try {
+                var sr = Parser.ReadStreamWithEncoding(stream, ProjectState.LanguageVersion);
```
`ReadStreamWithEncoding` wraps `stream` with `PartiallyReadStream` and then `stream` is set to `null`, so when `TextReader` is disposed, underlying stream will stay alive.

AlexanderSher commented on Python/Product/Analysis/LanguageServer/ParseQueue.cs:
```diff
@@ -0,0 +1,198 @@
+                    // We need to cancel immediately if we're running on the same thread
+                    // This can happen if the task completes before we call ContinueWith,
+                    // though the RunContinuationsAsynchronously should protect against this too
+                    if (Thread.CurrentThread == callingThread) {
```
Delegate passed to `ContinueWith` may be executed in the caller thread only if `TaskContinuationOptions.ExecuteSynchronously` is explicitly passed to `ContinueWith`. Otherwise, task will be queued into scheduler. Otherwise, you can safely call `ContinueWith` for completed task.

zooba replied:
The debugger disagrees with you, though with the RunContinuationsAsynchronously it seems to work. Still, better to have the safety net in there than deal with bug reports later on (which I can guarantee will be vague performance related issues)

AlexanderSher replied:
> The debugger disagrees with you

That's something. Any specific repro?

AlexanderSher replied:
Ok, I was able to get an exception when `Thread.CurrentThread == callingThread`. There is no `ParseQueue.Parse` in the call stack, just continuation executed in the same `ThreadPool` thread the task with `ParseQueue.Parse` was executed before. This is normal - `ThreadPoolTaskScheduler` doesn't start new thread for each task, they are reused.

I don't have a repro, but here is what I think has happened:
1. `previousTask.Status == TaskStatus.Running` was true
2. `ParseQueue.Parse` has returned `res` (which wasn't canceled at that moment)
3. `res` was saved into `_parsing` and returned from `ParseQueue.Enqueue`
4. Since task is canceled, `ParseWorker` and `ProjectEntry.UpdateTree` won't be executed
5. `AnalysisQueue.Worker` will wait for `_workEvent` forever. This is the reasons (or at least one of the reasons) for the hanging @huguesv has found.

AlexanderSher replied:
So, if you have seen a continuation delegate and `ParseQueue.Parse` being in the same call stack without `TaskContinuationOptions.ExecuteSynchronously` explicitly specified, I believe it should be filed as a bug in Task implementation, cause such behavior is unexpected (synchronously executed continuation is an ad-hoc scenario).

If it is just the thread ids that match, then I think this method should be reduced to 
```cs
        private Task<IAnalysisCookie> Parse(IDocument doc, PythonLanguageVersion languageVersion, Task<IAnalysisCookie> previousTask) 
            => previousTask?.ContinueWith(t => ParseWorker(doc, languageVersion), CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default)
               ?? Task.Factory.StartNew(() => ParseWorker(doc, languageVersion), CancellationToken.None, TaskCreationOptions.None, TaskScheduler.Default);
```

`TaskScheduler.Default` ensures that task won't be inlined because of some unexpected logic in the `TaskScheduler.Current`. You can use `TaskCreationOptions.RunContinuationsAsynchronously` and `TaskContinuationOptions.RunContinuationsAsynchronously` to ensure that **all** continuations of these tasks are executed as separate thread pool work items.

AlexanderSher commented on Python/Product/Analysis/Intellisense/AnalysisQueue.cs:
```diff
@@ -64,12 +63,42 @@ sealed class AnalysisQueue : IDisposable {
+
+        public event EventHandler<UnhandledExceptionEventArgs> UnhandledException;
+
+        public TaskScheduler Scheduler { get; private set; }
```
Is it still in use?

AlexanderSher commented on Python/Product/Analysis/Intellisense/AnalysisQueue.cs:
```diff
@@ -145,10 +174,15 @@ sealed class AnalysisQueue : IDisposable {
             }
         }
 
+        public void WaitForActivity(int millisecondsTimeout) {
```
Method isn't used anywhere, and as a result, `_activityEvent` isn't used as well.

AlexanderSher commented on Python/Product/Analysis/Intellisense/LineInfo.cs:
```diff
@@ -0,0 +1,81 @@
+        /// 0-based line number
+        /// </summary>
+        public int LineNo;
+        public NewLineKind LineBreak;
```
Why these fields aren't read-only?

AlexanderSher replied:
Also, I believe this type can be a struct.

zooba replied:
No idea, I didn't write this class. But you're right, it does only seem to be treated as an integral value, so it can be a struct.

AlexanderSher commented on Python/Product/Analysis/Intellisense/LinePreservingCodeReplacer.cs:
```diff
@@ -0,0 +1,159 @@
+
+            int curOldLine = 0;
+            for (int curNewLine = 0; curOldLine < _oldLines.Count && curNewLine < _newLines.Count; curOldLine++) {
+                if (GetOldText(curOldLine) == GetNewText(curNewLine)) {
```
You can do
```cs
                var oldLine = _oldLines[curOldLine];
                var newLine = _newLines[curNewLine];
                if (oldLine.Length == newLine.Length && string.CompareOrdinal(_oldCode, oldLine.Start, _newCode, newLine.Start, newLine.Length) == 0) {
                    curNewLine++;
                    continue;
                }
```
That's a bit more code, but it is faster and no string allocation for equal lines

zooba replied:
I'm hesitant to change this code right now - it can be changed in its own PR (this PR has just moved it)

AlexanderSher commented on Python/Product/Analysis/Intellisense/LinePreservingCodeReplacer.cs:
```diff
@@ -0,0 +1,159 @@
+                }
+            }
+
+            if (curOldLine < _oldLines.Count) {
```
Do you need to handle remaining `_newLines` as well?

AlexanderSher commented on Python/Product/Analysis/Intellisense/LinePreservingCodeReplacer.cs:
```diff
@@ -0,0 +1,159 @@
+                bool replaced = false;
+                // replace starts on this line, figure out where it ends...
+                int startNewLine = curNewLine;
+                for (curNewLine += 1; curNewLine < _newLines.Count; curNewLine++) {
```
If `curOldLine` line is removed, there will be no matching in the `oldLineMapping`, and as a result single line deletion will be treated as a replacement of the whole remaining text.

AlexanderSher commented on Python/Product/Analysis/Intellisense/LinePreservingCodeReplacer.cs:
```diff
@@ -0,0 +1,159 @@
+                bool replaced = false;
+                // replace starts on this line, figure out where it ends...
+                int startNewLine = curNewLine;
+                for (curNewLine += 1; curNewLine < _newLines.Count; curNewLine++) {
```
Sorry, my original comment was incorrect.

But there are still some cases when a simple change can result in a replacement of a big chunk.
For instance, there is a subsequence of `n+1` lines `A..NA`, and user adds a new line to the beginning of it: `ZA..NA`. Loop will start with `curNewLine == 1`, `GetNewText(curNewLine)` will return line `A`, `lines` will contain `0` and `n`, and as a result `A..NA` will be replaced with `ZA`. The remaining comparison will most likely end up with full replacement.

I'm not sure how often such thing may happen and how critical is it at all.

zooba replied:
File a new issue for this - I'm not going to start messing with this algorithm right now.

AlexanderSher replied:
Ok. https://github.com/Microsoft/PTVS/issues/3566

AlexanderSher commented on Python/Product/Analysis/Intellisense/ProjectEntryMap.cs:
```diff
@@ -47,13 +48,13 @@ sealed class ProjectEntryMap : IEnumerable<KeyValuePair<string, IProjectEntry>>
                 id = _ids.Count;
             }
-            _projectFiles[filename] = node;
+            _projectFiles[new Uri(filename, UriKind.RelativeOrAbsolute)] = node;
```
Is it ok to have both relative and absolute paths in the same map?

zooba replied:
This entire class is unused now - I'll delete it.

AlexanderSher commented on Python/Product/Analysis/LanguageServer/CallbackEventArgs.cs:
```diff
@@ -0,0 +1,56 @@
+            _task = task;
+        }
+
+        public T? @params { get; set; }
```
`Params`?

zooba replied:
Matching the spec here, using the prefix for keywords. If anything, I'd change it to `params_`, but it's actually easier to type with the `@` at the start.

AlexanderSher replied:
Ok

AlexanderSher commented on Python/Product/Analysis/LanguageServer/DiagnosticsErrorSink.cs:
```diff
@@ -0,0 +1,95 @@
+        private readonly List<Diagnostic> _diagnostics;
+        private readonly IReadOnlyList<KeyValuePair<string, DiagnosticSeverity>> _taskCommentMap;
+
+        public DiagnosticsErrorSink(string source, List<Diagnostic> diagnostics, IReadOnlyDictionary<string, DiagnosticSeverity> taskCommentMap = null) {
```
`DiagnosticsErrorSink` is called in two places and both times `diagnostics = new List<Diagnostic>()`, so you can drop this parameter.

zooba replied:
That actually changes the semantics away from what I want this class to be. I'll change the argument to `Action<Diagnostic>` and remove the (unused) Diagnostics property.

AlexanderSher commented on Python/Product/Analysis/LanguageServer/Messages.cs:
```diff
@@ -0,0 +1,302 @@
+    }
+
+    public struct InitializeResult {
+        public ServerCapabilities? capabilities;
```
`struct` with mutable field. `capabilities` can be made `readonly` and initialized from constructor

zooba replied:
These are the public interface for the LSP implementation and I want them to be as close as possible to JSON objects for ease of mapping.

AlexanderSher replied:
Ok, got it.

AlexanderSher commented on Python/Product/Analysis/LanguageServer/Messages.cs:
```diff
@@ -0,0 +1,302 @@
+using System.Threading.Tasks;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    public struct InitializeParams {
```
struct with mutable fields

AlexanderSher commented on Python/Product/Analysis/LanguageServer/ServerBase.cs:
```diff
@@ -0,0 +1,232 @@
+            Volatile.Read(ref _lock)?.Cancel();
+        }
+
+        public virtual async Task<MessageActionItem?> ShowMessageRequest(ShowMessageRequestParams @params) {
```
Not used anywhere

zooba replied:
Part of the LSP spec

AlexanderSher commented on Python/Product/Analysis/LanguageServer/ServerBase.cs:
```diff
@@ -0,0 +1,232 @@
+        protected void Telemetry(TelemetryEventArgs e) => OnTelemetry?.Invoke(this, e);
+
+        public event EventHandler<RegisterCapabilityEventArgs> OnRegisterCapability;
+        protected Task RegisterCapability(RegistrationParams @params) {
```
`RegisterCapability`, `UnregisterCapability` and related events aren't used. Is it for the future functionality?

zooba replied:
Part of the LSP spec

AlexanderSher commented on Python/Product/Analysis/LanguageServer/VolatileCounter.cs:
```diff
@@ -0,0 +1,96 @@
+
+        private sealed class DecrementOnDispose : IDisposable {
+            public VolatileCounter _self;
+            public void Dispose() => _self.Decrement();
```
You can ensure that `Decrement` is called only once: `public void Dispose() => Interlocked.Exchange(ref _self, null)?.Decrement();`

AlexanderSher commented on Python/Product/Analysis/LanguageServer/VolatileCounter.cs:
```diff
@@ -0,0 +1,96 @@
+        /// Decrements the counter
+        /// </summary>
+        /// <returns></returns>
+        public bool Decrement() {
```
Public increment/decrement is usually an error-prone. Unless there is a memory-critical code path where additional IDisposable instance can make a problem, I think it is safer to make both these methods private.

zooba replied:
So review the usage carefully (it's needed for ParseQueue when we increment before scheduling the worker and decrement on failure or completion from a different thread). In this case, I believe it's simpler and clearer than passing around a disposable.

AlexanderSher commented on Python/Product/Analysis/LanguageServer/Server.cs:
```diff
@@ -0,0 +1,841 @@
+                    _pendingParse[doc] = counter = new VolatileCounter();
+                    // Automatically remove counter from the dictionary when it
+                    // reaches zero.
+                    counter.WaitForZeroAsync().ContinueWith(t => {
```
For new `VolatileCounter`, `WaitForZeroAsync` will return completed task because `IsZero == true`.
Because of the lock, `ContinueWith` will be executed right after counter is incremented, `if (counter.IsZero` check will fail and `doc` will be referenced forever.

zooba replied:
Ah, that's why they weren't being removed. Good catch!

AlexanderSher commented on Python/Product/Analysis/LanguageServer/VolatileCounter.cs:
```diff
@@ -0,0 +1,96 @@
+            if (IsZero) {
+                return Task.CompletedTask;
+            }
+            var tcs = Volatile.Read(ref _waitForZero);
```
If while current thread is on this line another thread decrements `_count`, it is possible that `_waitForZero.SetResuls` will never be called.

AlexanderSher commented on Python/Product/Analysis/LanguageServer/ServerBase.cs:
```diff
@@ -0,0 +1,232 @@
+            public readonly CancellationTokenSource CancellationTokenSource;
+
+            public CancellationToken Token => CancellationTokenSource.Token;
+            public void Cancel() => CancellationTokenSource.Cancel();
```
If `Cancel` is called after `Dispose`, there will be `ObjectDisposedException`

zooba replied:
Yes. You shouldn't call `CancelRequest` outside of an `AllowRequestCancellation` state or your object has been disposed.

AlexanderSher replied:
Ok

AlexanderSher commented on Python/Product/Analysis/LanguageServer/Server.cs:
```diff
@@ -0,0 +1,841 @@
+            }
+        }
+
+        private async void EnqueueItem(IDocument doc, AnalysisPriority priority = AnalysisPriority.Normal) {
```
`async void`

zooba replied:
Correct. This is fire-and-forget by design.

AlexanderSher commented on Python/Product/PythonTools/PythonTools/Intellisense/ProjectAnalyzer.cs:
```diff
@@ -993,7 +1025,53 @@ SafePipeHandle stdInClientHandle
+                }
+            ).ConfigureAwait(false);
+
+            return FinishAnalyze(
```
Should `FinishAnalyze` be called in `finally` block?

zooba replied:
No. This is a straight refactor splitting one method into two.

AlexanderSher commented on Python/Product/PythonTools/PythonTools/Intellisense/BufferParser.cs:
```diff
@@ -294,48 +259,86 @@ IEnumerable<ITextSnapshot> snapshots
-                if (res != null) {
-                    Debug.Assert(res.failed != true);
+                if (res?.failed ?? false) {
+                    Interlocked.Decrement(ref analyzer._parsePending);
```
Do we need a `finally` block somewhere here to ensure that `_parsePending` is decremented in case of exception?

zooba replied:
Any exception coming out of the Send call is irrecoverable.

AlexanderSher replied:
Ok

AlexanderSher commented on Python/Product/PythonTools/PythonTools/Editor/PythonTextBufferInfo.cs:
```diff
@@ -288,6 +308,31 @@ sealed class PythonTextBufferInfo {
+            if (entry != null) {
+                tcs.TrySetResult(entry);
+            } else if (cancellationToken.CanBeCanceled) {
+                cancellationToken.Register(() => tcs.TrySetResult(null));
```
`TrySetCanceled`?

zooba replied:
By design. You need to check for null anyway, so no point forcing all the callers to also check for exceptions and treat them identically to null.

AlexanderSher replied:
That's a bit unusual. For every awaited async method that accepts `cancellationToken`, caller expects that the code below canceled await won't be executed, no matter how deep in the call stack the actual handler of cancellation is placed. 
Here it is actually not true. While immediate caller can check for `null`, other methods in call stack has no idea about it.
Right now, none of the callers provide real `cancellationToken`, but it may change in future.

AlexanderSher commented on Python/Product/PythonTools/PythonTools/PythonAnalysisClassifier.cs:
```diff
@@ -99,10 +98,14 @@ private class ClassificationComparer : IEqualityComparer<AP.AnalysisClassificati
             }
 
-            public int GetHashCode(AP.AnalysisClassification obj) => obj.start;
+            public int GetHashCode(AP.AnalysisClassification obj) => obj.startLine << 8 + obj.startColumn;
```
Why `endLine`, `endColumn` and `type` are skipped?

zooba replied:
Because it's just a hash code, and the start location should be unique anyway.

AlexanderSher commented on Python/Product/PythonTools/PythonTools/OutliningTaggerProvider.cs:
```diff
@@ -171,30 +168,19 @@ internal class OutliningTagger : ITagger<IOutliningRegionTag>, IPythonTextBuffer
-                    if (headerIndex < 0) {
-                        headerIndex = snapshot.GetLineFromPosition(start).End.Position;
+                    SnapshotPoint hintEnd = end;
+                    if (start.GetContainingLine().LineNumber + 5 < hintEnd.GetContainingLine().LineNumber) {
```
Why `5`?

zooba replied:
Ask the original author. I suspect it's to prevent unnecessarily large tooltips.

AlexanderSher commented:
>             _server.OnPublishDiagnostics += OnPublishDiagnostics;

These events aren't unsubscribe in `Dispose`

---
Refers to: Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs:69 in 60f3aca. [](commit_id = 60f3aca16f217cbb3d867961b4f1f5db26375501, deletion_comment = False)

AlexanderSher commented:
>                     (oldExtension as IDisposable)?.Dispose();

`oldExtension` can be disposed outside of the critical section. Not sure how critical is it.

---
Refers to: Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs:270 in 60f3aca. [](commit_id = 60f3aca16f217cbb3d867961b4f1f5db26375501, deletion_comment = False)

huguesv commented:
I tested it for another hour or so.

Feature-wise, completions seem to work as expected, including when importing packages defined in the project.

Unfortunately, completions don't work for very long. Within 5 to 10 mins of editing, all IntelliSense stops working. That's if I'm not running under the debugger, because if I do that I usually get stuck in an infinite wait at some point. I guess if I'm not debugging that's when IntelliSense stop working.

![image](https://user-images.githubusercontent.com/1696845/35021252-e92581ce-fae4-11e7-8d43-c4087d864b12.png)

I've been trying to find what causes it, but there's no simple instruction for it. Sometimes it's after I type really fast. Sometimes it's after I switch to a different file in the editor then come back. One time I added a unit test, ran it then after that wasn't getting completions. Reloading the solution gets things working again.

AlexanderSher commented on Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs:
```diff
@@ -1,2381 +1,2139 @@
-                            doc = param.Documentation,
-                            type = param.Type,
-                            variables = param.Variables?.Select(MakeReference).Where(r => r != null).ToArray()
-                  
```
These events aren't unsubscribed in `Dispose`

zooba replied:
Doesn't matter, neither object lives beyond the disposal.

AlexanderSher commented on Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs:
```diff
@@ -1,2381 +1,2139 @@
-                            doc = param.Documentation,
-                            type = param.Type,
-                            variables = param.Variables?.Select(MakeReference).Where(r => r != null).ToArray()
-                  
```
`async void`

zooba replied:
We're already crashing by the time we get here.

AlexanderSher commented on Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs:
```diff
@@ -1,2381 +1,2139 @@
-                            doc = param.Documentation,
-                            type = param.Type,
-                            variables = param.Variables?.Select(MakeReference).Where(r => r != null).ToArray()
-                  
```
`oldExtension` can be disposed outside of the critical section. Not sure how critical is it.

zooba commented:
@huguesv Can you enable the connection logs for `Analyzer` and `ProjectAnalyzer` and when you see this, send the logs to me?

AlexanderSher commented:
>             if (entry.Analysis != null) {

`GetPythonEntry` can return null, so here we can get `NRE`.

---
Refers to: Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs:1430 in 60f3aca. [](commit_id = 60f3aca16f217cbb3d867961b4f1f5db26375501, deletion_comment = False)

AlexanderSher commented:
>             if (entry.Analysis != null) {

`entry` isn't checked for null

---
Refers to: Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs:1431 in ada8e2a. [](commit_id = ada8e2a4a3182866a8c6a2b73672c5cb2e8ef5ec, deletion_comment = False)

AlexanderSher commented:
>             var tree = entry.Tree;

`entry` isn't checked for `null`

---
Refers to: Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs:2076 in ada8e2a. [](commit_id = ada8e2a4a3182866a8c6a2b73672c5cb2e8ef5ec, deletion_comment = False)

AlexanderSher commented:
>             var tree = entry.Tree;

`tree` isn't used later

---
Refers to: Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs:2076 in ada8e2a. [](commit_id = ada8e2a4a3182866a8c6a2b73672c5cb2e8ef5ec, deletion_comment = False)

AlexanderSher commented:
>             if (entry.Analysis != null) {

`entry` isn't checked for `null`

---
Refers to: Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs:1431 in ada8e2a. [](commit_id = ada8e2a4a3182866a8c6a2b73672c5cb2e8ef5ec, deletion_comment = False)
